### PR TITLE
get interval from redis/ES + data cleanup in MT + pass on step to graphite

### PR DESCRIPTION
**Pull request by Dieterbe**

### CODE REVIEW

**Comment by Dieterbe** [comment]
@woodsaj i'm happy with the first commit, been testing it in devstack, the 2nd commit has only been tested through unit tests so far, haven't played with it or tested performance yet. but this shows the direction I think we should take for https://github.com/raintank/graphite-api/issues/8 : metric-tank can fix up the data and graphite-raintank doesn't have to do any of that stuff anymore. thoughts?

**Comment by woodsaj** [comment]
@Dieterbe  Do you think we should just move nsq_metrics_to_elastic into metric-tank?

**Comment by Dieterbe** [comment]
I've also been thinking about this.
The main reason that nsq_metrics_to_{kairos,elasticsearch} were designed as separate consumers was so that each would have their own channel and queue so that ES or kairosdb downtimes could be handled separately, i.e. kairosdb downtime would not prevent ES updates from propagating, and ES downtime would not prevent data being added into kairosdb. 
If both were handled from the same consumer it doesn't seem sensible to send an ack if only one of the submit to ES or tsdb worked, so one would hold the other up.

Bringing both responsabilities into MT would make it simpler to run, especially for people who just want to run MT (and cassandra+ES) and expect things to work by sending their metrics into MT, perhaps via a graphite tcp input instead of having to worry about a separate metadata indexer service.

one thing that is different now though, is that "tsdb ingest" always succeeds because it's in process. it's only the external ES that is a liability. long term (once we get rid of the python graphite stack and do everything in go) we may want to have the primary index in memory if we can feasibly implement persistence and replication. 

for now we could treat ES outages as rare exceptions and don't store the values in the chunkserver when ES is down so that we don't have the problems around acking but half-consumed messages. but that puts more pressure on upgrade procedures etc.

**Comment by woodsaj** [comment]
What if we just treat ES as an index'ing service and not the primary store for index data, and instead persist metric data to cassandra.  ES being down and not receiving data is not a problem as when ES comes back up the index will be populated as new metrics are ingested.  While ES is down, we cant query from graphite anyway.

**Comment by Dieterbe** [comment]
TODO: make sure all targets requested in same request have same step.

**Comment by Dieterbe** [comment]
just added some more commits with misc fixes.

**Comment by Dieterbe** [comment]
just pushed 2 WIP commits. i'm working on a solver that can find the optimal step for all targets by scoring various solutions and punishing for cassandra lookups and runtime consolidation. it needs some mork work though.

**Comment by Dieterbe** [comment]
so the multi-target step solver should work now, at least based on all the unit tests.  and i also went ahead and did the metrics_to_elasticsearch deprecation, see also https://github.com/raintank/raintank-docker/pull/51

unfortunately i've been having issues with devstack today, metric-tank is spewing errors like
```
2015/12/21 09:44:45 [metricdef.go:235 GetMetricDefinition()] [E] elasticsearch query failed. 2015-12-21 09:44:45.319853618 +0000 UTC: Error [ElasticsearchIllegalArgumentException[No feature for name [metric_index]]] Status [400] [400]
```
so i haven't been able to do my usual playing with zoom levels and verifying data returned looks proper.  no idea where this is coming from so suddenly.
I don't have enery anymore today to figure this out, hopefully @woodsaj knows what's up.

**Comment by Dieterbe** [comment]
@woodsaj can you please review this, hopefully it's good to merge (or good enough with any needed fixes forthcoming).  i still haven't been able to find out what the ES errors are about (see above comment, but now on line 255 in metricdef.go) but i'm getting them consistently and querying `http://localhost:9200/metric/_search?q=*:*` works fine and returns plenty of metrics with ` "_type": "metric_index",` set. also this started happening after i moved the nsq_metrics_to_elasticsearch logic into NMT i think.

File: metric_tank/http.go [hunk 1]
```diff
@@ -127,10 +128,25 @@ func Get(w http.ResponseWriter, req *http.Request, metaCache *MetaCache, aggSett
 			http.Error(w, "unrecognized consolidation function", http.StatusBadRequest)
 			return
 		}
-		log.Debug("===================================")
 		req := NewReq(id, fromUnix, toUnix, minDataPoints, maxDataPoints, consolidator)
+		reqs = append(reqs, req)
```
Comment (woodsaj) [inline metric_tank/http.go]: This is wrong.  reqs is initialized with a size of len(targets).  By appending here, you are leaving len(targets) empty items in reqs.

This then leads to metricdef.GetMetricDefinition() being called with an empty ("") metric id, which in turn leads to `2015/12/21 09:44:45 [metricdef.go:235 GetMetricDefinition()] [E] elasticsearch query failed. 2015-12-21 09:44:45.319853618 +0000 UTC: Error [ElasticsearchIllegalArgumentException[No feature for name [metric_index]]] Status [400] [400]` you have been seeing.

**Comment by woodsaj** [comment]
There is still a few changes needed here.  Mostly around the consolidation functions. They currently dont handle NaN values well, which are now being added by the fix() function. eg, Min consolidation of NaN points results in the value being set to math.MaxFloat64

**Comment by woodsaj** [comment]
Still seeing issues with some of the consolidation functions.

When querying for 7days of `litmus.*.*.ping.avg` with maxDataPoints set to 20, i get a result of 132 for one of the points.  This is not possible as the values are either 0 or 1.

**Pull request merged**